quickwit-common/src/uri.rs

```diff
@@ -83,48 +90,103 @@ impl Uri {
 
                 path.replace_range(0..1, &home_dir_path);
             }
-
-            if !path.starts_with('/') {
+            if Path::new(&path).is_relative() {
                 let current_dir = env::current_dir().context(
                     "Failed to resolve current working directory: dir does not exist or \
                      insufficient permissions.",
                 )?;
                 path = current_dir.join(path).to_string_lossy().to_string();
             }
-
             path = normalize_path(Path::new(&path))
                 .to_string_lossy()
                 .to_string();
         }
-
         Ok(Self {
             uri: format!("{}{}{}", protocol, PROTOCOL_SEPARATOR, path),
             protocol_idx: protocol.len(),
         })
     }
 
-    /// Returns the URI extension.
+    /// Constructs a [`Uri`] from a properly formatted string `<protocol>://<path>` where `path` is
+    /// normalized. Use this method exclusively for trusted input.
+    pub fn new(uri: String) -> Self {
+        let protocol_idx = uri
+            .find(PROTOCOL_SEPARATOR)
+            .expect("URI lacks protocol separator. Use `Uri::new` exclusively for trusted input.");
+        Self { uri, protocol_idx }
+    }
+
+    /// Returns the extension of the URI.
     pub fn extension(&self) -> Option<Extension> {
         Path::new(&self.uri)
             .extension()
             .and_then(OsStr::to_str)
             .and_then(Extension::maybe_new)
     }
 
-    /// Returns the uri protocol.
+    /// Returns the URI as a string slice.
+    pub fn as_str(&self) -> &str {
+        &self.uri
+    }
+
+    /// Returns the protocol of the URI.
     pub fn protocol(&self) -> &str {
         &self.uri[..self.protocol_idx]
     }
 
-    /// Returns the file path from the uri.
-    /// Useful only for `file://` protocol Uri.
+    /// Returns the file path of the URI.
+    /// Applies only to `file://` URIs.
     pub fn filepath(&self) -> Option<&Path> {
-        if self.protocol() == "file" {
+        if self.protocol() == FILE_PROTOCOL {
             self.uri.strip_prefix("file://").map(Path::new)
         } else {
             None
         }
     }
+
+    /// Consumes the [`Uri`] struct and returns the normalized URI as a string.
+    pub fn into_string(self) -> String {
+        self.uri
+    }
+
+    /// Creates a new [`Uri`] with `path` adjoined to `self`.
+    /// Fails if `path` is absolute, it replaces the current path.
```

Comment:
Is `/// Fails if `path` is absolute.` not sufficient for the docs? Do you want to justify why it fails? In this case I would write `Fails if `path` is absolute as we don't want to replace the current path`?

Reply:
Yes, I don't know what happened here, seems like an unfortunate copy and paste :D
Will fix.